### PR TITLE
speed

### PR DESCRIPTION
**Pull request by Quoc-Wolf**

### CODE REVIEW

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 1]
```diff
@@ -8,23 +13,101 @@
 import asiantech.quocnp.pronuciation_alarm.R;
 import asiantech.quocnp.pronuciation_alarm.fragment.MainFragment;
 import asiantech.quocnp.pronuciation_alarm.fragment.MainFragment_;
+import asiantech.quocnp.pronuciation_alarm.untils.SettingStore;
+import lombok.Getter;
 
 @EActivity(R.layout.activity_home)
 public class HomeActivity extends BaseActivity {
+    private final static int TIME_DELAY_EXIT_APP = 2000;
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:22]: static should comment

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 2]
```diff
@@ -8,23 +13,101 @@
...
 
     private static final String TAG = HomeActivity.class.getSimpleName();
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:24]: static should comment

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 3]
```diff
@@ -8,23 +13,101 @@
...
     //This is Fragment Main
     @FragmentByTag("MainFragment")
     MainFragment mMainFragment;
 
+    boolean doubleBackToExitPressedOnce = false;
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: private boolean mDoubleBackToExitPressedOnce = false;

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 1 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:22]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 4]
```diff
@@ -8,23 +13,101 @@
...
+
     @Override
     void afterView() {
-        //=================================
         if (mMainFragment == null) {
             mMainFragment = MainFragment_.builder().build();
         }
         FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: not understand ft

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 2 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:24]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 5]
```diff
@@ -8,23 +13,101 @@
...
         ft.add(R.id.frame_layout, mMainFragment, "MainFragment").commit();
+        SettingStore.setIdFragment(this, currentFragment.MAIN_FRAGMENT.getValueEnum());
+    }
+
+    //check when click button back of Devices
+    @Override
+    public boolean onKeyDown(int keyCode, KeyEvent event) {
+
+        if (keyCode == KeyEvent.KEYCODE_BACK) {
+            if (SettingStore.getIdFragment(this) == currentFragment.MAIN_FRAGMENT.getValueEnum()) {
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:54]: if many if,should have comment

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 3 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 6]
```diff
@@ -8,23 +13,101 @@
...
-        //=================================
         if (mMainFragment == null) {
             mMainFragment = MainFragment_.builder().build();
         }
         FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
         ft.add(R.id.frame_layout, mMainFragment, "MainFragment").commit();
+        SettingStore.setIdFragment(this, currentFragment.MAIN_FRAGMENT.getValueEnum());
+    }
+
+    //check when click button back of Devices
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 7]
```diff
@@ -8,23 +13,101 @@
...
+        }
+        return true;
+    }
+
+
+    /**
+     * This is method check twice click button back of Devices and exit app
+     */
+
+    @Override
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:69]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 4 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 8]
```diff
@@ -8,23 +13,101 @@
...
+    public void onBackPressed() {
+        if (doubleBackToExitPressedOnce) {
+            super.onBackPressed();
+            return;
+        }
+
+        this.doubleBackToExitPressedOnce = true;
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: delete this

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 9]
```diff
@@ -8,23 +13,101 @@
...
+                .setMessage("You are doing test? and you want to back Home now ?")
+                .setPositiveButton("OK", new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int which) {
+                        mMainFragment.popChildFragment();
+                        SettingStore.setIdFragment(HomeActivity.this, currentFragment.MAIN_FRAGMENT.getValueEnum());
+                    }
+                })
+                .setNegativeButton("No", new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int which) {
+                        // do nothing
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:101]: delete if not use

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java [hunk 10]
```diff
@@ -8,23 +13,101 @@
...
+     * This is method enum set current Fragment
+     */
+    public enum currentFragment {
+        MAIN_FRAGMENT(0),
+        CHILD_FRAGMENT(1);
+
+        @Getter
+        int valueEnum;
+
+        // Enum constructor
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 11]
```diff
@@ -31,6 +42,7 @@
 @EFragment(R.layout.fragment_child)
 public class ChildFragment extends BaseFragment {
     private static final String TAG = ChildFragment.class.getSimpleName();
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:47]: commnent

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 12]
```diff
@@ -31,6 +42,7 @@
...
+    private static final int REQUEST_CODE_RESULT = 777;
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:49]: comment

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 13]
```diff
@@ -39,8 +51,13 @@
     @ViewById(R.id.imgOfWord)
     ImageView mImgOfWord;
     List<Vocabulary> mVocabularies;
+    Context mContext;
     //count size of list vocabularies
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: count is index of list not size

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 5 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:54]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 14]
```diff
@@ -77,10 +102,21 @@ private void setNext() {
             mCount++;
         } else {
             //back Frangment Main
-            popChildFragment();
+            mHandler.sendMessage(new Message());
         }
     }
 
+    private void handlerBackPopChilFragment() {
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:113]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 6 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 7 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:69]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 8 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 15]
```diff
@@ -135,4 +171,99 @@ private String loadJSONFromAsset() {
         return json;
 
     }
+
+    public void speak() {
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:182]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 16]
```diff
@@ -135,4 +171,99 @@ private String loadJSONFromAsset() {
...
+                .getPackageName());
+        mIntent.putExtra(RecognizerIntent.EXTRA_ONLY_RETURN_LANGUAGE_PREFERENCE,
+                "true");
+        mIntent.putExtra("android.speech.extra.EXTRA_ADDITIONAL_LANGUAGES",
+                new String[]{"en"});
+        mIntent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1);
+        startActivityForResult(mIntent, REQUEST_CODE_RESULT);
+    }
+
+    // Su kien nhan lai ket qua
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: java doc

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 17]
```diff
@@ -135,4 +171,99 @@ private String loadJSONFromAsset() {
...
+    private void showDialogWrong(String textResult) {
+        new AlertDialog.Builder(getMainActivity())
+                .setTitle("Result...")
+                .setMessage("You are wrong! \n" + "This is a word you say:" + textResult)
+                .setPositiveButton("Try again", new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int which) {
+                        speak();
+                    }
+                })
+                .setNegativeButton("Exit", new DialogInterface.OnClickListener() {
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: not use should delete

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java [hunk 18]
```diff
@@ -135,4 +171,99 @@ private String loadJSONFromAsset() {
...
+                if (!textMatchList.isEmpty()) {
+                    if (textMatchList.get(0).contains("search")) {
+                        String searchQuery = textMatchList.get(0).replace(
+                                "search", " ");
+                        Intent search = new Intent(Intent.ACTION_WEB_SEARCH);
+                        search.putExtra(SearchManager.QUERY, searchQuery);
+                        startActivity(search);
+                    } else {
+                        // Hien thi ket qua
+                        Log.i("FOX", "text--" + mTvNameWord.getText() + "//speed--" + textMatchList.get(0));
```
Comment (NguyenCongBi) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: Log.i(TAG,...);

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 9 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java:101]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java (hunk 10 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/activity/HomeActivity.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 11 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:47]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 12 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:49]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 13 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 14 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:113]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 15 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java:182]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 16 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: has fix

File: app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java (hunk 18 above)
Comment (Quoc-Wolf) [inline app/src/main/java/asiantech/quocnp/pronuciation_alarm/fragment/ChildFragment.java]: has fix

**Pull request merged**